python/ray/train/callbacks/results_preprocessors/aggregate/aggregate_preprocessor.py

```diff
@@ -0,0 +1,150 @@
+import logging
+from typing import Dict, List, Optional
+
+import numpy as np
+
+from ray.util.debug import log_once
+from ray.util.annotations import DeveloperAPI
+from ray.train.callbacks.results_preprocessors.preprocessor import ResultsPreprocessor
+from ray.train.callbacks.results_preprocessors.aggregate.aggregate_fn import (
+    AggregateFn,
+    Average,
+    Max,
+    WeightedAverage,
+)
+from ray.train.callbacks.results_preprocessors.aggregate.aggregate_utils import (
+    VALID_AGGREGATE_TYPES,
+    _get_metrics_from_results,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class AggregateResultsPreprocessor(ResultsPreprocessor):
+    """A preprocessor that aggregates training metrics from all workers.
+
+    Args:
+        aggregation_fn (AggregateFn):
+            An aggregation method that performs the aggregation on results.
+        keys (Optional[List[str]]):
+            A list of keys reported in results to be aggregated. Keys should
+            be saved using ``train.report()``. If a key is invalid or not
+            reported, it will be ignored.
+    """
+
+    def __init__(self, aggregation_fn: AggregateFn, keys: Optional[List[str]] = None):
+        self.aggregate_fn = aggregation_fn
+        self.keys = keys
+
+    def preprocess(self, results: List[Dict]) -> List[Dict]:
+        """Aggregate results before sending them to callbacks.
+
+        Args:
+            results (List[Dict]): A list of results from all workers. The metrics
+                specified in ``keys`` will be averaged according by ``aggregation_fn``.
+                Non-numerical values will be ignored.
+        Returns:
+            An updated results list with aggregated results.
+        """
+        results = [] if results is None else results
+        if len(results) == 0:
+            return results
+
+        self.aggregate_fn.prepare(results)
+        reported_metrics = {key for result in results for key in result.keys()}
+
+        if self.keys is None:
+            valid_keys = []
+            for metric in reported_metrics:
+                if all(
+                    isinstance(result.get(metric, np.nan), VALID_AGGREGATE_TYPES)
+                    for result in results
+                ):
+                    valid_keys.append(metric)
+                elif log_once(metric):
+                    logger.warning(
+                        f"`{metric}` value type is not "
+                        f"one of {VALID_AGGREGATE_TYPES}, so it is ignored. "
+                    )
+
+            self.keys = valid_keys
```

Comment:
Do we need to set `self.keys` here? Can we just pass in the keys directly into `_get_metrics_from_results`?

Reply:
Hi @amogkam, thank you very much for your feedback! I gather a detailed explanation for your comments in this single thread.

The challenging situation with `_get_metrics/weights_from_results` and `valid_keys` is because
> we allow users to input keys (to be aggregated) as `None`. If so, the default behavior is to aggregate all valid keys ([suggested](https://github.com/ray-project/ray/pull/22099#discussion_r810552642) by Matthew). 

To answer your question:
> I don't think this is doing what you intend it to do? Do you want to return None if the key is not reported by _any_ single worker? Or if it's not included in _all_ workers?

the intention is to ignore the key (two cases) (1) if no worker reports it, or (2) the values returned from all workers are invalid, for example, strings. However, it is allowed to have `np.nan` for some workers but not all, in which the metric is _not_ skipped. (Please let me know if this sounds reasonable to you.)

It seems the behavior of `_get_metrics/weights_from_results` and checking key validation are confusing. So I actually thinking of the following: Define two new functions, `_check_if_key_is_reported()` and `_check_if_value_is_valid()`. We use them to avoid duplicate code, i.e., 
```python
if self.keys is None:
    valid_keys = []
    for metric in reported_metrics:
        if _check_if_key_is_reported(metric, results):
            valid_keys.append(metric)
        elif log_once(metric):
            # log warning
    self.keys = valid_keys
```
and (similar for `_get_weights_from_results()`)
```python
def _get_metrics_from_results(key: str, results: List[Dict]):
    warning_message = None
    if not _check_if_key_is_reported(key, results):
        warning_message = (
            f"`{key}` is not reported from workers, so it is ignored. "
            "Please make sure that it is saved using `train.report()`."
        )
    elif not _check_if_value_is_valid(key, results):
        warning_message = (
            f"`{key}` value type (`{type(values[0])}`) is not valid, "
            "so it is ignored. "
            f"Make sure that its type is one of {VALID_AGGREGATE_TYPES}. "
        )

    if warning_message:
        logger.warning(warning_message)
        return None

    return [result.get(key, np.nan) for result in results]
```

For `log_once()`, I guess I am thinking too much. I imagined a case (now it seems very unlikely) where there is inconsistency between two epochs, e.g. the logging of two epochs failed due to different errors. So warning logging is kept not for once. But it seems I was overly concerned.

Will add other tests soon. Thanks again!

Reply:
Thanks for the detailed explanation here @jwyyy!

Yes the two helper functions sound like a great idea! But still, why would you need to call `_check_if_value_is_valid` twice? Isn't is sufficient to just call it only in `_get_metrics_from_results`. I put together a gist on what `preprocess` could look like: https://gist.github.com/amogkam/4b0957f7ee865c57736cf6ef1c4d1ccc, let me know if this works!

> the intention is to ignore the key (two cases) (1) if no worker reports it, or (2) the values returned from all workers are invalid, for example, strings. However, it is allowed to have np.nan for some workers but not all, in which the metric is not skipped. (Please let me know if this sounds reasonable to you.)

Yes this sounds good! I think I might have misunderstood one of your previous comments. So in that case, the metric is only aggregated for the workers that contain it, is that right? The behavior sounds good, but the comments still stand- can you provide more clarity on this behavior in the docstrings?

And also, is this the same behavior for the weight key? In that case, what is the behavior if some of the workers have the weight key and some don't?

> For log_once(), I guess I am thinking too much. I imagined a case (now it seems very unlikely) where there is inconsistency between two epochs, e.g. the logging of two epochs failed due to different errors. So warning logging is kept not for once. But it seems I was overly concerned.

Yeah this is a good point, definitely could happen, but as you said probably won't be common. I would optimize for avoiding too many spammy logs. And in fact, if it's possible to do one warning message for all the keys that are skipped, that might be best.

Reply:
@amogkam Thank you so much for your suggestions! They look great!

> And also, is this the same behavior for the weight key? In that case, what is the behavior if some of the workers have the weight key and some don't?

For weight keys, if they are missing or invalid, equal weight will be used. In the case when both metric and weight are valid, but some values are missing (as you pointed out, some workers may miss weights, and other may miss metrics), then only those that have both are aggregated. (Actually I realized that the current [implementation](https://github.com/ray-project/ray/pull/22099/files#diff-eb187d87055dc2cbe5a8259bc8d5851a2405d23282f059071cdaf7c82ebbcc48R77) doesn't really do this well, it should be 
```python
np.nansum(np.array(values) * self.weights / np.nansum(self.weights * (1 - np.isnan(values)) ))
```
) Please let me know if this idea sounds reasonable to you.

I will revise the PR and address all issues soon. Thanks a lot! (will write better docstring.)

Reply:
Yes this sgtm!